regression-tests/pure2-enum.cpp2

```diff
@@ -117,4 +119,10 @@ main: () = {
         is (cpp2::has_flags(f2))      = "includes all f2's flags ('cached' and 'current')";
         is _                          = "something else";
     } << "\n";
+
+    f_from_string := file_attributes::from_string("cached_and_current");
+    std::cout << "f_from_string is " << f_from_string.to_string() << "\n";
```

Comment:
It bothers me slightly that both of these:

```
file_attributes::to_string(file_attributes::from_string("cached_and_current"));
(file_attributes::cached | file_attributes::current).to_string()
```

result in this:

```
"(cached, current, cached_and_current)"
```

I'm wondering if a bitwise enum `to_string()` should be a single combined value if it exactly matches a combined value, and then if it doesn't, only include the bare flags.

That would make the first ones result in `"cached_and_current"`.  Then `(file_attributes::current | file:attributes.obsolete).to_string()` would be `(current, obsolete)`.

Also, should `to_string()` include `file_attributes::`?

Reply:
Thanks!

> It bothers me slightly that both of these:
> 
> ```
> file_attributes::to_string(file_attributes::from_string("cached_and_current"));
> (file_attributes::cached | file_attributes::current).to_string()
> ```
> 
> result in this:
> 
> ```
> "(cached, current, cached_and_current)"
> ```
> 
> I'm wondering if a bitwise enum `to_string()` should be a single combined value if it exactly matches a combined value, and then if it doesn't, only include the bare flags.

What I've been doing with `to_string` is emitting all of the named values that are matched. That way it's consistent with `has`.

> Also, should `to_string()` include `file_attributes::`?

`to_string` has a `this` parameter, `from_string` doesn't so it needs the qualification.

Reply:
>What I've been doing with to_string is emitting all of the named values that are matched. That way it's consistent with has.

My "nightmare scenario" is something like the creating an enum for the file permissions in `<sys/stat.h>`, and then doing `ALLPERMS.to_string()` and getting this:

```
"(S_ISUID, S_ISGID, S_ISTXT, S_ISVTX, S_IRWXU, S_IRUSR, S_IWUSR, S_IXUSR, 
S_IREAD, S_IWRITE, S_IEXEC, S_IRWXG, S_IRGRP, S_IWGRP, S_IXGRP, 
S_IRWXO, S_IROTH, S_IWOTH, S_IXOTH, ACCESSPERMS, ALLPERMS, 
DEFFILEMODE)"
```

>to_string has a this parameter, from_string doesn't so it needs the qualification.

It was just a thought, not for computer consumption, but more for human consumption.  I was thinking that the user always has to spell it as `file_attributes::current`, unless they're inside the class, so would it make sense for `to_string()` to produce that?  That would make `from_string()` more difficult, I guess.

Reply:
> Related to that, would it make sense to use | instead of , so that it makes actually usable as code?

👍🏻 I was thinking of this myself when I used `flag_enum`s a while ago.

Reply:
Thanks! I just pushed a commit to change `,` to `|` for flag enums.

Allowing qualification is doable but just makes reading and parsing harder. A poor person's way to make it a qualified string would be to prepend `flag_name::` and then replace every `| ` with `| flag_name::`... perhaps it's worth adding a `to_string_qualified` that just calls `to_string` and then does that, and a `from_string_qualified` that replaces all `flag_name::` with nothing and calls `from_string`...?

Reply:
Indeed, it could be a thing. Python for example has `__repr__` and `__str__` which cover both of these use-cases.

Reply:
I'm not sure if there are use cases here, just things that came to mind as I was reading it.  

Maybe the "a single combined value if it exactly matches a combined value, and then if it doesn't, only include the bare flags" would be a separate function and it could have a `qualified` option?